### PR TITLE
nix-universal-prefetch: init at 0.2.0

### PR DESCRIPTION
**Pull request by samueldr**
Upstream URL: https://github.com/samueldr/nix-universal-prefetch

###### Motivation for this change

Let's hopefully have users use this!

This software is intended to resolve once and for all the question "but I don't want to TOFU?" (which isn't even a question!) when end-users want to get the hash to use with a specific fetcher.

> For anyone confused by TOFU, one of the quick replies on IRC is:
> > To get a sha256 hash of a new source, you can use the Trust On First Use model: use probably-wrong hash (for example: 0000000000000000000000000000000000000000000000000000) then replace it with the correct hash Nix expected.

This, in addition, could be used with automated scripts to reduce hand-holding when the fetchers aren't simple underneath.

###### Things done

- ✔️ Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - ✔️ NixOS
   - ⬜ macOS
   - ⬜ other Linux distributions
- ⬜ Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- ✔️ Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- ✔️ Tested execution of all binary files (usually in `./result/bin/`)
- ⬜ Determined the impact on package closure size (by running `nix path-info -S` before and after)
- ✔️ Assured whether relevant documentation is up to date
- ✔️ Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Additionally, I ate my own dogfood when making this PR, and did **not** TOFU the hash, but instead ran:

 ```
nix-universal-prefetch fetchFromGitHub --owner samueldr --repo nix-universal-prefetch --rev v0.1.0
```

---

### CODE REVIEW

File: pkgs/tools/package-management/nix-universal-prefetch/default.nix [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+      --replace "/usr/bin/env nix-shell" "${ruby}/bin/ruby"
+  '';
+
+  meta = with stdenv.lib; {
+    description = "Uses nixpkgs fetchers to figure out hashes";
+    homepage = https://github.com/samueldr/nix-universal-prefetch;
+    license = licenses.mit;
+    maintainers = with maintainers; [ samueldr ];
+    platforms = platforms.linux ++ platforms.darwin;
+    inherit version;
```
[the remark this answers is not in the document]
Comment (samueldr) [reply pkgs/tools/package-management/nix-universal-prefetch/default.nix]: Frankly, you'd have to ask me past July, when I added `nix-top` as I copied its derivation and removed what wasn't required, while not really looking at `meta`. I must have had a brain fart when implementing that.

File: pkgs/tools/package-management/nix-universal-prefetch/default.nix [hunk 2]
```diff
@@ -0,0 +1,32 @@
+{ stdenv
+, fetchFromGitHub
+, ruby
+}:
+
+# No gems used, so mkDerivation is fine.
+stdenv.mkDerivation rec {
+  name = "nix-universal-prefetch-${version}";
```
Comment (msteen) [inline pkgs/tools/package-management/nix-universal-prefetch/default.nix]: Support for `pname` and `version` [has been added to `stdenv.mkDerivation`](https://github.com/NixOS/nixpkgs/blob/c8c53fcb1154999dabac350bebe12611a0f75024/pkgs/stdenv/generic/make-derivation.nix#L84) (it is not backported), so considering you already do add `version`, you should probably split `name` in `pname` as well.

**Comment by samueldr** [comment]
The update (0.2.0) adds support for nix 2.2's new error message. It also supports `builtins.fetchurl` which has a different message.

All this is now tested, though not tested in this build as the test is made using nix.

**Comment by danbst** [comment]
Another case when universal prefetcher would be better then url prefetcher https://github.com/NixOS/nixpkgs/commit/e36b4d6a857afb13840a5b4468b82a69eeca45d4

**Comment by danbst** [comment]
Though, to be honest, it is not true:

> Additionally, I ate my own dogfood when making this PR, and did not TOFU the hash, but instead ran:

This prefetcher is TOFU, but hidden from eyes. What would be extra awesome, if in next iteration:
- some support for **checking** TLS certificates is added to nixpkgs (I can think of a configuration parameter, which when passed to nixpkgs makes `fetch*` check certificates). Requested by @7c6f434c 
- this prefetcher then applies that config by default when prefetching things. When certificates check fail, it should fail as well and point to MITM attack page in docs
- then rename this to `nix-prefetch`
- autocomplete!

Then we can make this an "officially blessed" prefetcher and mention in docs.

**Comment by samueldr** [comment]
You're right, not not-TOFU since it's TOFU, but automated for you, but isn't `nix-prefetch` TOFU in that sense (in the current implementation)?

And right again, this would need the implementation details of *further checks* to be made in nixpkgs, since there is no introspection possible here as to making further checks.

[samueldr/nix-universal-prefetch#1](https://github.com/samueldr/nix-universal-prefetch/issues/1) is a tracking issue if you want to reference it in the future.

* * *

As for autocomplete, there's an annoying hard problem here: there is no way to introspect deep enough to show all arguments. Let's use `fetchFromGitHub` as an example, all of *its* arguments are discoverable, but it will pass-through some arguments to either `fetchurl` or `fetchgit`, depending on some conditions. What could be done, though, is to supplement the list of discovered arguments with a list of known arguments, either through making "dumb" passthrough arguments inside nixpkgs, or inside nix-universal-prefetch. (Implementation inside nixpkgs is probably preferrable.)

* * *

And thanks for the interest :).

**Comment by msteen** [comment]
I have been working on my own idea for `nix-prefetch`: https://github.com/msteen/nix-prefetch. I think I have managed to improve upon all aspects of `nix-universal-prefetch` and have still some more ideas on how to improve it further.

@danbst Could you give me more information about that TLS support, I like to implement your suggestion. Also with regards to auto complete, I have not implemented it yet, but I do have all the information available to make it work, even considering the problem mentioned by @samueldr. The only limitation being is that for fetchers (mine supports both pointing to fetchers and extending existing fetcher calls, like package sources), I would need to keep a list of valid bogus values, or else risk producing errors.

**Comment by danbst** [comment]
@msteen it is basically set here: https://github.com/NixOS/nix/blob/800cd55ab74fffe9a1d212c2b8caf4024ebd98b6/src/libstore/download.hh#L16

Pretty impressed, that there are now 2 implementations of same concept!

**Comment by 7c6f434c** [comment]
@msteen re: TLS: all (if I remember correctly; in any case, most) of the Nixpkgs fetchers intended for direct use in packages do not verify certificates; the assumption is that if hash is given, why bother how the content was obtained — and size leaks do indeed kill all or almost all privacy when downloading tarballs (size leaks can even damage privacy of encrypted VoIP, and popular tarballs on a single server are known to attackers in advance).

With TOFU that means zero MITM protection.

The main benefit is probably that an old Nixpkgs checkout without fresh certificates can still build something.

I think, given the popularity of manual TOFU which is not going to disappear overnight, we should convert all the fetchers to opt-out certificate checking: the default is to check, but there are both a fetcher parameter and a global option to skip verification (the first is useful for packages with known-non-trivial setups, like special-purposes CA certificates; the second is useful in case Nixpkgs checkout is too old)

**Pull request merged**